jetty-io/src/main/java/org/eclipse/jetty/io/ssl/SslConnection.java

```diff
@@ -480,134 +393,183 @@ protected WriteFlusher getWriteFlusher()
         @Override
         protected void onIncompleteFlush()
         {
-            // This means that the decrypted endpoint write method was called and not
-            // all data could be wrapped. So either we need to write some encrypted data,
-            // OR if we are handshaking we need to read some encrypted data OR
-            // if neither then we should just try the flush again.
-            boolean try_again = false;
-            boolean write = false;
-            boolean need_fill_interest = false;
-            synchronized (DecryptedEndPoint.this)
+            ByteBuffer write = null;
+            boolean fill_interest = false;
+            synchronized(_decryptedEndPoint)
             {
                 if (LOG.isDebugEnabled())
-                    LOG.debug("onIncompleteFlush {}", SslConnection.this);
-                // If we have pending output data,
-                if (BufferUtil.hasContent(_encryptedOutput))
-                {
-                    // write it
-                    _cannotAcceptMoreAppDataToFlush = true;
-                    write = true;
-                }
-                // If we are handshaking and need to read,
-                else if (_sslEngine.getHandshakeStatus() == HandshakeStatus.NEED_UNWRAP)
-                {
-                    // check if we are actually read blocked in order to write
-                    _flushRequiresFillToProgress = true;
-                    need_fill_interest = !SslConnection.this.isFillInterested();
-                }
-                else
+                    LOG.debug("onIncompleteFlush {} {}", SslConnection.this, BufferUtil.toDetailString(_encryptedOutput));
+                
+                switch(_flushState)
                 {
-                    // TODO This should not be required and we should be able to do all retries within flush
-                    // We can get here because the WriteFlusher might not see progress
-                    // when it has just flushed the encrypted data, but not consumed anymore
-                    // of the application buffers.  This is mostly avoided by another iteration
-                    // within DecryptedEndPoint flush(), but this still occurs sometime on some
-                    // tests on some systems??? More investigation is needed!
-                    try_again = true;
+                    case NEEDS_WRITE:
+                        _flushState = FlushState.WRITING;
+                        write = _encryptedOutput==null?BufferUtil.EMPTY_BUFFER:_encryptedOutput;
+                        break;
+
+                    case NEEDS_UNWRAP:
+                        if (_sslEngine.getHandshakeStatus()==HandshakeStatus.NEED_UNWRAP)
```

Comment:
I can't see how this is right. The state should not be modified meanwhile, as we're holding the lock. If it can be modified concurrently, then certainly this check is subject to a race, so not good.